compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/generators/CallAndReferenceGenerator.kt

```diff
@@ -361,14 +361,35 @@ class CallAndReferenceGenerator(
         }.applyCallArguments(annotationCall, annotationMode = true)
     }
 
-    fun convertToGetObject(qualifier: FirResolvedQualifier): IrExpression {
-        return convertToGetObject(qualifier, callableReferenceMode = false)!!
+    internal fun convertToGetObject(qualifier: FirResolvedQualifier): IrExpression {
+        return convertToGetObject(qualifier, null)!!
     }
 
-    internal fun convertToGetObject(qualifier: FirResolvedQualifier, callableReferenceMode: Boolean): IrExpression? {
+    internal fun convertToGetObject(
+        qualifier: FirResolvedQualifier,
+        callableReferenceAccess: FirCallableReferenceAccess? = null
+    ): IrExpression? {
         val classSymbol = (qualifier.typeRef.coneType as? ConeClassLikeType)?.lookupTag?.toSymbol(session)
-        if (callableReferenceMode && classSymbol is FirRegularClassSymbol) {
-            if (classSymbol.classId != qualifier.classId) {
+        if (callableReferenceAccess != null && classSymbol is FirRegularClassSymbol) {
+            val classIdMatched = classSymbol.classId == qualifier.classId
+            val companionClassIdMatched = classSymbol.fir.isCompanion && classSymbol.classId.outerClassId == qualifier.classId
+            val lastClassIfCompanion = conversionScope.lastClass()?.takeIf { it.isCompanion }
+            val callable =
+                ((callableReferenceAccess.calleeReference as FirResolvedNamedReference).resolvedSymbol as FirCallableSymbol<*>)
+            val companionMemberReferenceThatNeedsDispatchReceiver = companionClassIdMatched &&
```

Comment:
I must confess all this code looks like a hack, sorry. Will try to invent something better...

Reply:
Thank you for the all simplifications!